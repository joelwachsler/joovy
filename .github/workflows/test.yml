name: Test
on:
  push:
    tags:
      - '!**'
jobs:
  test:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Run tests
        run: yarn test

  lint:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Lint
        run: yarn lint
