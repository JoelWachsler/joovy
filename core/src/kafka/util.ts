import { Kafka } from 'kafkajs'
import { defer, from, mapTo, mergeMap, Observable, tap } from 'rxjs'
import initConfig from '../config'
import logger from '../logger'
import { KMessage } from './kafka'

let kafka: Kafka

/**
 * Internal implementation detail, do not use.
 */
export const createKafkaClient = () => {
  if (!kafka) {
    const kafkaConfig = initConfig().kafka

    kafka = new Kafka({
      clientId: kafkaConfig.clientId,
      brokers: kafkaConfig.brokers,
    })
  }

  return kafka
}

type CreateConsumerArgs = {
  topic: string
  groupId: string
}

export const createConsumer = <T extends KMessage = KMessage>(args: CreateConsumerArgs) => {
  const { groupId, topic } = args

  const create = defer(() => {
    logger.info(`Creating consumer with the following args: ${JSON.stringify(args)}`)
    const consumer = createKafkaClient().consumer({ groupId })

    return from(consumer.connect()).pipe(mapTo(consumer))
  })

  const connectToTopic = create.pipe(mergeMap(consumer => {
    logger.info(`Subscribing to topic: ${topic}...`)
    return from(consumer.subscribe({ topic, fromBeginning: true })).pipe(
      tap(() => logger.info(`Subscribed to topic: ${topic}!`)),
      mapTo(consumer),
    )
  }))

  return connectToTopic.pipe(mergeMap(consumer => {
    return new Observable<T>(subscribe => {
      consumer.run({
        async eachMessage({ message, topic }) {
          const msgAsString = message.value?.toString()
          if (!msgAsString) {
            logger.error(`Invalid message: ${message} for topic: ${topic}`)
            return
          }

          logger.info(`New incoming message: ${JSON.stringify(msgAsString)} on topic: ${topic}`)

          subscribe.next(JSON.parse(msgAsString))
        },
      })
    })
  }))
}