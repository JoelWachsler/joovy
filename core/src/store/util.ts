import { defer } from 'rxjs'
import JEvent from '../jevent/JEvent'

const messageStore = new Map<string, Map<string, unknown>>()

export const getOrCreateObjectStore = (id: string) => {
  let store = messageStore.get(id)
  if (!store) {
    store = new Map()
    messageStore.set(id, store)
  }

  return store
}

export const removeObjectStore = (event: JEvent) => {
  return defer(() => {
    const channelId = event.message.channelId
    messageStore.delete(channelId)

    return event.result({ store: { event: 'removed', id: channelId } })
  })
}
