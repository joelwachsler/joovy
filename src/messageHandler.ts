import { mergeMap, Observable, of } from 'rxjs'
import { handle } from './commands/command'
import JEvent from './JEvent'

export const handleMessage = (event$: Observable<JEvent>): Observable<JEvent> => {
  return event$.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return of(event.withResult({ ignored: `${message.content} was sent by a bot` }))
      } else if (!message.content.startsWith('/')) {
        return of(event.withResult({ ignored: `${message.content} does not start with a slash` }))
      } else {
        return handle(event)
      }
    }),
  )
}
