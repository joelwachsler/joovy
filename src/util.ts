
export namespace Util {

  const apply = <T>(t: T) => {
  }
}
