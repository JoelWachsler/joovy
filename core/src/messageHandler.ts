import { Client } from 'discord.js'
import { catchError, filter, mergeMap, Observable } from 'rxjs'
import { handle } from './commands/command'
import { errorHandler } from './errorHandler'
import JEvent, { fromMessageKey } from './jevent/JEvent'
import { EmptyResult, Result } from './jevent/Result'
import { createConsumer, createProducer, KMessage, Topics } from './kafka/kafka'
import { logResult } from './resultLogger'

export const sendMessageEvent = (event: Observable<JEvent>): Observable<Result> => {
  return createProducer().pipe(mergeMap(producer => {
    return event.pipe(mergeMap(event => {
      return producer.send({
        event,
        message: createMessageEvent(event),
        topic: Topics.NewMessage,
      })
    }))
  }))
}

export const handleMessageEvents = (client: Client) => {
  const messageHandling = createConsumer<MessageEvent>({ groupId: 'joovy-message-handler', topic: Topics.NewMessage }).pipe(
    mergeMap(msg => fromMessageKey(client, msg.meta.messageKey)),
    handleMessage,
  )

  logResult('handleMessageEvents', messageHandling)
}

export const handleMessage = (event: Observable<JEvent>): Observable<Result> => {
  return event.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return event.result({ ignored: `${message.content} was sent by a bot` })
      } else if (!message.content.startsWith('/')) {
        return event.result({ ignored: `${message.content} does not start with a slash` })
      } else {
        return handle(event).pipe(
          filter(r => !(r instanceof EmptyResult)),
          catchError(err => errorHandler(event, err)),
        )
      }
    }),
  )
}

const createMessageEvent = (event: JEvent): MessageEvent => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'message-event',
      version: '1',
    },
    content: event.message.content,
  }
}

type MessageEvent = KMessage<'message-event', '1'> & { content: string }
