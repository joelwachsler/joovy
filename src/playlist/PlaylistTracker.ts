import { randomUUID } from 'crypto'
import { concat, defaultIfEmpty, map, mapTo, mergeAll, mergeMap, Observable, of } from 'rxjs'
import JEvent from '../jevent/JEvent'
import { StringStore } from '../store/Store'
import { Playlist } from './Playlist'

export const trackPlaylist = (event: JEvent, playlist: Playlist) => {
  return event.store.persistentString.pipe(
    map(store => createRepositories(store)),
    mergeMap(repos => {
      const playlistId = randomUUID()
      const { channelId } = event.message

      const createPlaylist = repos.channel.getOrCreate(event, channelId).pipe(
        mergeMap(channel => {
          channel.playlists = [...channel.playlists, playlistId]
          return repos.channel.put(channelId, channel)
        }),
      )

      const tracking = concat(
        createPlaylist,
        playlist.internalCurrentQueue.pipe(
          mergeMap(currentQueue => {
            return repos.playlist.getOrCreate(event, playlistId).pipe(
              mergeMap(queue => {
                queue.tracks = currentQueue.map(track => {
                  return {
                    version: '1',
                    author: {
                      version: '1',
                      id: track.author.id,
                      username: track.author.username,
                    },
                    name: track.name,
                    link: track.link,
                  }
                })

                return repos.playlist.put(playlistId, queue)
              }),
            )
          }),
        ),
      )

      return tracking.pipe(
        mergeMap(e => event.complexResult({ result: { playlist: 'trackingEvent' }, item: e })),
      )
    }),
  )
}

abstract class Repository<Entity extends Version, ID extends string = string> {
  constructor(private store: StringStore) { }

  get(id: ID): Observable<Entity> {
    return this.store.get(id).pipe(map(e => JSON.stringify(e) as unknown as Entity))
  }

  put(id: ID, entity: Entity): Observable<Entity> {
    return this.store.put(id, JSON.stringify(entity)).pipe(mapTo(entity))
  }

  getOrCreate(event: JEvent, id: ID): Observable<Entity> {
    return this.get(id).pipe(
      map(e => of(e)),
      defaultIfEmpty(this.put(id, this.defaultEntity(event))),
      mergeAll(),
    )
  }

  abstract defaultEntity(event: JEvent): Entity
}

const createRepositories = (store: StringStore) => {
  return {
    channel: new class extends Repository<ChannelV1> {
      defaultEntity(): ChannelV1 {
        return {
          version: '1',
          playlists: [],
        }
      }
    }(store),
    playlist: new class extends Repository<PlaylistV1> {
      defaultEntity(event: JEvent): PlaylistV1 {
        return {
          version: '1',
          date: event.timestamp,
          tracks: [],
        }
      }
    }(store),
  }
}

interface ChannelV1 extends Version<'1'> {
  playlists: PlaylistIdentifier[]
}

type PlaylistIdentifier = string

interface PlaylistV1 extends Version<'1'> {
  date: number
  tracks: TrackV1[]
}

interface TrackV1 extends Version<'1'> {
  author: AuthorV1
  name: string
  link: string
}

interface AuthorV1 extends Version<'1'> {
  id: string
  username: string
}

interface Version<Version = string> {
  version: Version
}
