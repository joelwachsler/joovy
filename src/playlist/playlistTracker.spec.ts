import { map, Observable } from 'rxjs'
import { handleMessage } from '../messageHandler'
import { createTestEvent, hot, sandbox } from '../test/util'
import { TRACKING_EVENT } from './PlaylistTracker'

test('should create tracker for playlist', () => {
  const play = createTestEvent({ content: '/play test' })

  const messages = handleTrackingMessages(hot('a', { a: play }))
  expect(messages).toMatchSnapshot()
})

test('should track multiple tracks in the same playlist', () => {
  const play = createTestEvent({ content: '/play test' })
  const play2 = createTestEvent({ content: '/play test2' })

  const messages = handleTrackingMessages(hot('ab', { a: play, b: play2 }))
  expect(messages).toMatchSnapshot()
})

test('should differentiate tracking between playlists', () => {
  const play = createTestEvent({ content: '/play test' })
  const disconnect = createTestEvent({ content: '/disconnect' })

  const messages = handleTrackingMessages(hot('aba', { a: play, b: disconnect }))
  expect(messages).toMatchSnapshot()
})

const handleTrackingMessages = (source: Observable<any>) => sandbox.getMessages(handleMessage(source).pipe(
  map(msg => {
    // include the actual item when matching tracking events
    if (msg.result === TRACKING_EVENT) {
      return msg
    } else {
      return msg.result
    }
  }),
))
