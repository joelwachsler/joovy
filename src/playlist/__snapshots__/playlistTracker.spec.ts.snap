// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create tracker for playlist 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "playlists": Array [
            "totally-random-uuid-0",
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test.com",
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
              "version": "1",
            },
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should differentiate tracking between playlists 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "playlists": Array [
            "totally-random-uuid-0",
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test.com",
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
              "version": "1",
            },
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/disconnect",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "disconnected",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Bye!\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "playlists": Array [
            "totally-random-uuid-0",
            "totally-random-uuid-1",
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test.com",
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
              "version": "1",
            },
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 7,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 7,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should track multiple tracks in the same playlist 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "playlists": Array [
            "totally-random-uuid-0",
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test.com",
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
              "version": "1",
            },
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test.com",
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
              "version": "1",
            },
            Object {
              "author": Object {
                "id": "testAuthorId",
                "username": "testuser",
                "version": "1",
              },
              "link": "https://test2.com",
              "name": "[test2 (1:10)](https://test2.com) [<@testAuthorId>]",
              "version": "1",
            },
          ],
          "version": "1",
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;
